evaluation/what-to-support.md

```diff
@@ -0,0 +1,152 @@
+        // ...
+    }
+}
+
+trait Iterator {
+    type Item;
+    // `map` is always `const`.
+    //
+    // This is the case as `map` only creates the `Map`
+    // struct but does not eagerly use `f`.
+    const fn map<B, F>(self, f: F) -> Map<Self, F>
+    where
+        Self: Sized,
+        F: FnMut(Self::Item) -> B,
+    {
+        Map::new(self, f)
+    }
+}
+
+pub struct Map<I, F> {
+    iter: I,
+    f: F,
+}
+
+impl<B, I, F, const<C>> const<C> Iterator for Map<I, F>
+where
+    I: const<C> Iterator,
+    F: const<C> FnMut(I::Item) -> B,
+{
+    type Item = B;
+    // ...
+}
+```
+
+## how does `impl const<C> Trait` work
+
+`const` trait impls are intended to mean that **all** functions of that trait are const. But, how const?
+
+As seen in the `Iterator::map` example up above, simply adding `const<C>` to all trait bounds of trait methods is too restrictive. The same holds for a lot of other iterator adaptors.
+
+A similar issue is dropping generic parameters. Looking at the following example:
+```rust
+trait Trait {
+    fn takes_self(self);
+
+    fn doesnt_take_self();
+
+    fn has_other_param<T>();
+
+    fn uses_other_param<T>(x: T);
+}
+```
+
+What would a `const` impl of that trait look like? The most permissive option for implementors would be:
+```rust
+impl<T, const<C>> const<C> Trait for T {
+    fn takes_self(self)
+    where
+        T: const<C> Destruct,
+    {}
+
+    fn doesnt_take_self()
+    where
+        T: const<C> Destruct,
+    {}
+
+    fn has_other_param<U>()
+    where
+        T: const<C> Destruct,
+        U: const<C> Destruct,
+    {}
+
+    fn uses_other_param<U>(x: U)
+    where
+        T: const<C> Destruct,
+        U: const<C> Destruct,
+    {}
+}
+```
+While the impl which can be used in the most contexts would have no `const<C> Destruct` bounds at all. Any automated way of figuring this out would sometimes be incorrect, as functions with equal signatures can still have different requirements.
+
+Considering that traits apparently have to opt-in to being `const`-able, due to backwards compatability of default methods, we could require the trait itself to declare its bounds when const, which may look like the following:
+```rust
+const<C> trait Trait {
+    // Sidenote: the `X: const<C> Destruct` for generic parameters
+    // of the trait could instead be moved to the super traits or
+    // the impl instead. This is probably the more sensible approach
+    // for most traits.
+    fn takes_self(self)
+    where
+        Self: const<C> Destruct;
+
+    // No bounds.
+    fn doesnt_take_self();
+
+    // Not visible in the method signature,
+    // but maybe its expected that this function creates
+    // and drops an instance of the `Self` type, so the
+    // trait author assumes that `Self` can be dropped.
+    fn has_other_param<T>();
+    where
+        Self: const<C> Destruct;
+
+    fn uses_other_param<T>(x: T)
+    where
+        T: const<C> Destruct;
+}
+```
+
+## multi `const` effect fun
```

Comment:
Can you explain in one sentence what the _goal_ of this example is? From what I'm reading it seems to be something like this:

> This example shows two arguments, whose `const`ness is independent of one another. But in order for the function `merge` to be `const`, both parameters need to be `const`.

Separating the _why_ from the _how_ seems particularly important when talking about use problem definitions.

Reply:
Note: I think we talked about it two weeks ago, and I believe we concluded that assigning individual effects to arguments which _don't_ affect their outer function is a use case we need to support?

Though I guess what this says is that the constness of `C1` and `C2` is independent of one another, but both are needed to mark `merge` as const. Do we have other alternatives of encoding this? Can this be assumed to always be the case? How common would the desire to encode these semantics be? (I assume: always?)

Reply:
> Though I guess what this says is that the constness of C1 and C2 is independent of one another, but both are needed to mark merge as const. 

jup

> Do we have other alternatives of encoding this?

:shrug: i thought of one, added it to the doc, though i already dislike that one xd

> Can this be assumed to always be the case?

the alternative only needs one constness parameter

> How common would the desire to encode these semantics be? 

common enough to ideally be supported :grin: